fvm/transactionStorageLimiter.go

```diff
@@ -3,42 +3,96 @@ package fvm
 import (
 	"fmt"
 
+	"github.com/opentracing/opentracing-go"
+
+	"github.com/onflow/cadence"
 	"github.com/onflow/cadence/runtime/common"
 
 	errors "github.com/onflow/flow-go/fvm/errors"
 	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/trace"
 )
 
 type TransactionStorageLimiter struct{}
 
-func NewTransactionStorageLimiter() *TransactionStorageLimiter {
-	return &TransactionStorageLimiter{}
+func NewTransactionStorageLimiter() TransactionStorageLimiter {
+	return TransactionStorageLimiter{}
 }
 
-func (d *TransactionStorageLimiter) CheckLimits(
+// TransactionStorageLimiterScriptArgumentBatchSize is the number of accounts to check at once.
+// Using to many arguments at once might cause problems during InvokeAccountsStorageCapacity.
+const TransactionStorageLimiterScriptArgumentBatchSize = 100
+
+func (d TransactionStorageLimiter) CheckLimits(
 	env Environment,
 	addresses []flow.Address,
 ) error {
 	if !env.Context().LimitAccountStorage {
 		return nil
 	}
 
-	// iterating through a map in a non-deterministic order! Do not exit the loop early.
-	for _, address := range addresses {
-		commonAddress := common.Address(address)
+	var span opentracing.Span
+	if te, ok := env.(*TransactionEnv); ok && te.isTraceable() {
+		span = te.ctx.Tracer.StartSpanFromParent(te.traceSpan, trace.FVMTransactionStorageUsedCheck)
+		defer span.Finish()
+	}
 
-		capacity, err := env.GetStorageCapacity(commonAddress)
+	commonAddresses := make([]common.Address, len(addresses))
+	usages := make([]uint64, len(commonAddresses))
+	for i, address := range addresses {
+		c := common.Address(address)
+		commonAddresses[i] = c
+		u, err := env.GetStorageUsed(c)
 		if err != nil {
 			return fmt.Errorf("storage limit check failed: %w", err)
 		}
+		usages[i] = u
+	}
 
-		usage, err := env.GetStorageUsed(commonAddress)
+	for start := 0; start < len(addresses); start += TransactionStorageLimiterScriptArgumentBatchSize {
+		end := start + TransactionStorageLimiterScriptArgumentBatchSize
+		if end > len(addresses) {
+			end = len(addresses)
+		}
+
+		err := d.batchCheckLimits(
+			env,
+			commonAddresses[start:end],
+			usages[start:end],
+			span,
+		)
```

Comment:
any concern here that we break it into smaller batches ?

Reply:
I am somewhat concerned that sending too many arguments at once might inflate the memory costs. But the limit I set may be a bit low on second though... should be more around 10_000 or something

I just don't want it to be limitless really.

Reply:
but isn't it anyway it might fail because of the memory ?

Reply:
I did some more thinking. 

My fear was that you could create an attack where you would touch so many accounts that getting the storage for all of them at once would OOM. 

But I think this problem might not be possible due to the fact that touching accounts requires computation, and even if it is, batching might not be the right approach to solve it. 

I have deleted the batching part to reduce complexity.